applications/plugins/SofaPython/Binding_Base.cpp

```diff
@@ -35,84 +36,353 @@ using namespace sofa::core::objectmodel;
+    const char* dataHelp="";
+    PyObject* dataValue = nullptr;
+
+    char *dataName; // The desired name is provided using regular args ...
+    if(!PyArg_ParseTuple(args, "s",&dataName))
+    {
+        msg_error("SosaPython") << "No name provided for the new Data";
+        return nullptr;
+    }
+
+    // ... the other values are passed as kwargs
+//    if (kw && PyDict_Size(kw)>0)
+//    {
+//        PyObject* keys = PyDict_Keys(kw);
+//        PyObject* values = PyDict_Values(kw);
+//        for (int i=0; i<PyDict_Size(kw); i++)
+//        {
+//            PyObject *key = PyList_GetItem(keys,i); // no need to use py_decref for these (using will actually cause segfault)
+//            PyObject *value = PyList_GetItem(values,i);
+//            std::string KeyStr = PyString_AsString(key);
+
+//            if (KeyStr.compare("datatype")==0)
+//            {
+//                dataRawType = getStringCopy(PyString_AsString(value));
+//            }
+//            if (KeyStr.compare("help")==0)
+//            {
+//                dataHelp = getStringCopy(PyString_AsString(value));
+//            }
+//            if (KeyStr.compare("dataclass")==0)
+//            {
+//                dataClass = getStringCopy(PyString_AsString(value));
+//            }
+//            if (KeyStr.compare("value")==0)
+//            {
+//                dataValue = value;
+//            }
+//        }
+//        Py_DecRef(keys); // TODO (Stefan Escaida): wtf is happening here??
+//        Py_DecRef(values);
+//    }
+
+    if(!PyDict_Check(kw))
+    {
+        msg_error("SofaPython") << "Could not parse arguments for adding Data";
+        return nullptr;
+    }
+
+    PyObject * tmp;
+    tmp = PyDict_GetItemString(kw,"datatype");
+    if (!(tmp==nullptr)){
+        dataRawType = getStringCopy(PyString_AsString(tmp));
+    }
+
+    tmp = PyDict_GetItemString(kw,"helptxt");
+    if (!(tmp==nullptr)){
+        dataHelp = getStringCopy(PyString_AsString(tmp));
+    }
+
+    tmp = PyDict_GetItemString(kw,"dataclass");
+    if (!(tmp==nullptr)){
+        dataClass= getStringCopy(PyString_AsString(tmp));
+    }
+
+    tmp = PyDict_GetItemString(kw,"value");
+    if (!(tmp==nullptr)){
+        dataValue = tmp;
+        Py_IncRef(dataValue); // call to Py_GetItemString doesn't increment the ref count, but we want to hold on to it for a while ...
+    }
+
+    if (dataRawType[0]==0) // We cannot construct without a type
+    {
+        msg_error("SofaPython") << "No type provided for Data, cannot construct/add";
+        return nullptr;
+    }
+
+    BaseData* bd = nullptr ;
+//    bool OldSchoolParsing = false; //TODO (Stefan Escaida 29.01.2018): should a direct parsing mechanism Python-->Sofa be implemented? (instead of going through str)
+    if(dataRawType[0] == 's'){
+        Data<std::string>* t = new Data<std::string>() ;
+        t = new(t) Data<std::string>(obj->initData(t, std::string(""), dataName, dataHelp)) ;
+        bd = t;        
+    }
```

Comment:
A factory would be beter :)
Something like : BaseData* bf = factory(dataRawType[0]);

Reply:
Here is an example of a factory with sofa : https://github.com/SofaDefrost/sofa/blob/sofaCommunication/applications/plugins/Communication/components/serverCommunicationOSC.inl#L50

Reply:
Ok, thanks! I will look into it.